sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -79,16 +79,16 @@ case class Filter(condition: Expression, child: SparkPlan)
 
   // Split out all the IsNotNulls from condition.
   private val (notNullPreds, otherPreds) = splitConjunctivePredicates(condition).partition {
-    case IsNotNull(a) if child.output.contains(a) => true
+    case IsNotNull(a) if a.references.subsetOf(child.outputSet) => true
```

Comment:
ah, you are right. With the `NullIntolerant` trait introduced in #11809, we can solve this easily. Wait for that PR to be ready then we can re-visit this. Thanks!